{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as patches\n",
    "import seaborn as sns\n",
    "import itertools\n",
    "import re\n",
    "import numpy as np\n",
    "sns.set_style('white')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# CLAMMS\n",
    "def readclamms(infiles, data):\n",
    "    for file in infiles:\n",
    "        with open(file) as fin:\n",
    "            next(fin)\n",
    "            for line in fin:\n",
    "                l = line.strip().split()\n",
    "                chrom,x,y,sample,cnv, cn = l[0], l[1], l[2], l[4], l[5], l[6]\n",
    "                if PLOT_TYPE not in sample:\n",
    "                    continue\n",
    "                if cnv == 'DUP':\n",
    "                    cnv = 'dup'\n",
    "                else:\n",
    "                    cnv = 'del'\n",
    "                if sample not in data:\n",
    "                    data[sample] = []\n",
    "                data[sample].append([chrom,int(x),int(y),cnv,int(cn), 'clamms'])\n",
    "\n",
    "# CODEX2\n",
    "def readcodex2(infiles, data):\n",
    "    for file in infiles:\n",
    "        with open(file) as fin:\n",
    "            next(fin)\n",
    "            for line in fin:\n",
    "                l = line.strip().split()\n",
    "                chrom,x,y,sample,cnv,cn = l[1], l[3], l[4], l[0], l[2], l[10]\n",
    "                if PLOT_TYPE not in sample:\n",
    "                    continue\n",
    "                if cnv == 'dup':\n",
    "                    pass\n",
    "                if sample not in data:\n",
    "                    data[sample] = []\n",
    "                data[sample].append([chrom,int(x),int(y),cnv,int(cn), 'codex2'])\n",
    "\n",
    "# CONTRA\n",
    "def readcontra(infiles, data):\n",
    "    for file in infiles:\n",
    "        with open(file) as fin:\n",
    "            next(fin)\n",
    "            for line in fin:\n",
    "                l = line.strip().split()\n",
    "                chrom,x,y,sample,cnv = l[4], l[5], l[6], l[0], l[14]\n",
    "                if PLOT_TYPE not in sample:\n",
    "                    continue\n",
    "                if cnv == 'gain':\n",
    "                    cnv, cn = 'dup', 3\n",
    "                else:\n",
    "                    cnv, cn = 'del', 1\n",
    "                if sample not in data:\n",
    "                    data[sample] = []\n",
    "                data[sample].append([chrom,int(x),int(y),cnv,cn, 'contra'])\n",
    "\n",
    "# EXCAVATOR2\n",
    "def readexcavator2(infiles, data):\n",
    "    for file in infiles:\n",
    "        with open(file) as fin:\n",
    "            next(fin)\n",
    "            for line in fin:\n",
    "                l = line.strip().split()\n",
    "                chrom,x,y,sample,cnv = l[1], l[2], l[3], l[0], int(l[7])\n",
    "                if PLOT_TYPE not in sample:\n",
    "                    continue\n",
    "                if cnv >= 1:\n",
    "                    cnv, cn = 'dup', 3\n",
    "                elif cnv == -1:\n",
    "                    cnv, cn = 'del', 1\n",
    "                else:\n",
    "                    cnv, cn = 'del', 0\n",
    "                if sample not in data:\n",
    "                    data[sample] = []\n",
    "                data[sample].append([int(x),int(y),cnv,cn, 'excavator2'])\n",
    "\n",
    "# EXOMEDEPTH\n",
    "def readexomedepth(infiles, data):\n",
    "    for file in infiles:\n",
    "        with open(file) as fin:\n",
    "            next(fin)\n",
    "            for line in fin:\n",
    "                l = line.strip().split()\n",
    "                chrom,x,y,sample,cnv = l[7].strip('\"'), l[5], l[6], l[0], l[3]\n",
    "                if cnv.strip('\"') == 'deletion':\n",
    "                    cnv, cn = 'del', 1\n",
    "                else:\n",
    "                    cnv, cn = 'dup', 3\n",
    "                if sample not in data:\n",
    "                    data[sample] = []\n",
    "                data[sample].append([chrom, int(x),int(y),cnv,cn, 'exomedepth'])\n",
    "\n",
    "# XHMM\n",
    "def readxhmm(infiles, data):\n",
    "    for file in infiles:\n",
    "        with open(file) as fin:\n",
    "            next(fin)\n",
    "            for line in fin:\n",
    "                l = line.strip().split()\n",
    "                chrom, xy, sample,cnv = l[4], l[2], l[0], l[1]\n",
    "                _,x,y = re.split('[:-]',xy)\n",
    "                if PLOT_TYPE not in sample:\n",
    "                    continue\n",
    "                if cnv == 'DEL':\n",
    "                    cnv, cn = 'del', 1\n",
    "                else:\n",
    "                    cnv, cn = 'dup', 3\n",
    "                if sample not in data:\n",
    "                    data[sample] = []\n",
    "                data[sample].append([chrom, int(x),int(y),cnv,cn, 'xhmm'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reading GFF3 file\n",
    "# Currently only reading entries for genes, ignoring alternative isoforms\n",
    "infile = '/home/harald/projects/ensembl/Homo_sapiens.GRCh38.95.chr.gff3'\n",
    "genesdb = {}\n",
    "exonsdb = {}\n",
    "with open(infile, 'r') as fin:\n",
    "    for line in fin:\n",
    "        if line.startswith('#'):\n",
    "            continue\n",
    "        l = line.strip().split()\n",
    "        if 'RNA' in l[2]:\n",
    "            for ann in l[8].split(';'):\n",
    "                key, value = ann.split('=')\n",
    "                if key == 'Name':\n",
    "                    name = value\n",
    "                elif key == 'ID':\n",
    "                    _, ts = value.split(':')\n",
    "            genesdb[ts] = name\n",
    "        elif l[2] == 'exon':\n",
    "            for ann in l[8].split(';'):\n",
    "                key, value = ann.split('=')\n",
    "                if key == 'Name':\n",
    "                    name = value\n",
    "                elif key == 'Parent':\n",
    "                    _, ts = value.split(':')\n",
    "            try:\n",
    "                exonsdb[name] = ['chr'+l[0], int(l[3]), int(l[4]),ts,genesdb[ts]]\n",
    "            except KeyError:\n",
    "                continue\n",
    "# exonsdb = {'ENSE00002234944': ['chr1', 11869, 12227, 'ENST00000456328', 'DDX11L1-202'], ..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reading collected results\n",
    "methods = {'codex2':1,'exomedepth':2,'xhmm':3}\n",
    "data = {}\n",
    "PLOT_TYPE = 'case'\n",
    "exomedepth_files = ['/home/harald/projects/somatic/test_samples/wes_cnvs/final_report_exomedepth.txt']\n",
    "excavator2_files = []\n",
    "codex2_files = ['/home/harald/projects/somatic/test_samples/wes_cnvs/final_report_codex2.txt']\n",
    "contra_files = []\n",
    "xhmm_files = ['/home/harald/projects/somatic/test_samples/wes_cnvs/final_report_xhmm.txt']\n",
    "clamms_files = []\n",
    "if len(exomedepth_files) > 0:\n",
    "    readexomedepth(exomedepth_files, data)\n",
    "if len(excavator2_files) > 0:\n",
    "    readexcavator2(excavator2_files, data)\n",
    "if len(contra_files) > 0:\n",
    "    readcontra(contra_files, data)\n",
    "if len(codex2_files) > 0:\n",
    "    readcodex2(codex2_files, data)\n",
    "if len(xhmm_files) > 0:\n",
    "    readxhmm(xhmm_files, data)\n",
    "if len(clamms_files) > 0:\n",
    "    readclamms(clamms_files, data)\n",
    "outfile = 'results2/cnv_{}.png'.format(PLOT_TYPE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "del_results = {}\n",
    "dup_results = {}\n",
    "for exon in exonsdb:\n",
    "    chrom, start, stop, ts, gene = exonsdb[exon]\n",
    "    for sample, records in data.items():\n",
    "        if sample not in del_results:\n",
    "            del_results[sample] = {}\n",
    "            dup_results[sample] = {}\n",
    "        for record in records:\n",
    "            #record : [chrom, int(x),int(y),cnv=['dup','del'],cn, 'xhmm']\n",
    "            if record[0] != chrom:\n",
    "                continue\n",
    "            if start < record[2] and record[1] < stop:\n",
    "                if record[3] == 'dup':\n",
    "                    dup_results[sample][exon] = dup_results[sample].get(exon, 0) + 1\n",
    "                else:\n",
    "                    del_results[sample][exon] = del_results[sample].get(exon, 0) + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('BTN-reportfile_duplicates_wes-cnv.txt', 'w') as fout:\n",
    "    fout.write('sample\\tgene\\ttranscript\\texon\\tchrom\\tstart\\tstop\\tcount\\n')\n",
    "    for sample, result in dup_results.items():\n",
    "        for exon in result:\n",
    "            chrom, start, stop, ts, gene = exonsdb[exon]\n",
    "            if chrom in ['chrY','chrX']:\n",
    "                continue\n",
    "            if result[exon] < 2:\n",
    "                continue\n",
    "            fout.write('{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\n'.format(sample,gene,ts,exon,chrom,start,stop,result[exon]))\n",
    "with open('BTN-reportfile_deletions_wes-cnv.txt', 'w') as fout:\n",
    "    fout.write('sample\\tgene\\ttranscript\\texon\\tchrom\\tstart\\tstop\\tcount\\n')\n",
    "    for sample, result in del_results.items():\n",
    "        for exon in result:\n",
    "            chrom, start, stop, ts, gene = exonsdb[exon]\n",
    "            if chrom in ['chrY','chrX']:\n",
    "                continue\n",
    "            if result[exon] < 2:\n",
    "                continue\n",
    "            fout.write('{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\n'.format(sample,gene,ts,exon,chrom,start,stop,result[exon]))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:datascience]",
   "language": "python",
   "name": "conda-env-datascience-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
